File: native/cocos/renderer/renderer/Camera.cpp
/****************************************************************************
 Copyright (c) 2018 Xiamen Yaji Software Co., Ltd.

 http://www.cocos2d-x.org
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/

#include "Camera.h"
#include "gfx/FrameBuffer.h"
#include "math/MathUtil.h"

RENDERER_BEGIN

Camera::Camera()
{
}

Camera::~Camera()
{
    RENDERER_SAFE_RELEASE(_framebuffer);
    RENDERER_SAFE_RELEASE(_node);
}

void Camera::setFrameBuffer(FrameBuffer* framebuffer)
{
    RENDERER_SAFE_RELEASE(_framebuffer);
    _framebuffer = framebuffer;
    RENDERER_SAFE_RETAIN(_framebuffer);
}

void Camera::setWorldMatrix(const Mat4& worldMatrix)
{
    Quaternion rotation;
    worldMatrix.decompose(nullptr, &rotation, &_worldPos);

    Mat4::createTranslation(_worldPos, &_worldRTInv);
    _worldRTInv.rotate(rotation);
    _worldRTInv.inverse();
}

void Camera::setColor(float r, float g, float b, float a)
{
    _color.set(r, g, b, a);
}

void Camera::setDepth(float depth)
{
    _depth = depth;
}

void Camera::setStencil(int stencil)
{
    _stencil = stencil;
}

void Camera::setClearFlags(uint8_t flags )
{
    _clearFlags = flags;
}

void Camera::setRect(float x, float y, float w, float h)
{
    _rect.set(x, y, w, h);
}

void Camera::setStages(const std::vector<std::string>& stages)
{
    _stages = stages;
}


void Camera::setNode(NodeProxy* node)
{
    if (_node != nullptr)
    {
        _node->release();
    }
    _node = node;
    if (_node != nullptr)
    {
        _node->retain();
    }
}

void Camera::calcMatrices(const int width, const int height)
{
    // view matrix
    //REFINE:
    _node->getWorldRT(&_worldRTInv);
    _matView.set(_worldRTInv.getInversed());

    // projecton matrix
    float aspect = (float)width / height;
    if (ProjectionType::PERSPECTIVE == _projection)
        Mat4::createPerspective(_fov / 3.1415926 * 180, aspect, _near, _far, &_matProj);
    else
    {
        float x = _orthoHeight * aspect;
        float y = _orthoHeight;
        Mat4::createOrthographicOffCenter(-x, x, -y, y, _near, _far, &_matProj);
    }

    // view projection
    Mat4::multiply(_matProj, _matView, &_matViewProj);
    _matInvViewProj.set(_matViewProj.getInversed());
}

void Camera::extractView(View& out, int width, int height)
{
    if (_framebuffer != nullptr) {
        width = _framebuffer->getWidth();
        height = _framebuffer->getHeight();
    }
    
    // rect
    out.rect.set(_rect.x * width,
                 _rect.y * height,
                 _rect.w * width,
                 _rect.h * height);
    
    // clear opts
    out.color.set(_color.r, _color.g, _color.b, _color.a);
    out.depth = _depth;
    out.stencil = _stencil;
    out.clearFlags = _clearFlags;

    calcMatrices(width, height);
    out.matView.set(_matView);
    out.matProj.set(_matProj);
    out.matViewProj.set(_matViewProj);
    out.matInvViewProj.set(_matInvViewProj);
    
    // stages & framebuffer
    out.stages = _stages;
    out.frameBuffer = _framebuffer;
    
    // culling mask
    out.cullingMask = _cullingMask;
    out.cullingByID = true;
}

Vec3& Camera::screenToWorld(Vec3& out, const Vec3& screenPos, int width, int height)
{
    calcMatrices(width, height);
    
    float cx = _rect.x * width;
    float cy = _rect.y * height;
    float cw = _rect.w * width;
    float ch = _rect.h * height;
    
    if (ProjectionType::PERSPECTIVE == _projection)
    {
        // Caculate screen pos in far clip plane.
        out.set((screenPos.x - cx) * 2.0f / cw - 1.0f,
                (screenPos.y - cy) * 2.0f / ch - 1.0f,
                1.0f);
        
        // Transform to world position.
        out.transformMat4(out, _matInvViewProj);
        
        _node->getWorldPosition(&_worldPos);
        _temp_v3.set(_worldPos);
        out = _temp_v3.lerp(out, MathUtil::lerp(_near / _far, 1, screenPos.z));
    }
    else
    {
        out.set((screenPos.x - cx) * 2.0f / cw - 1.0f,
                (screenPos.y - cy) * 2.0f / ch - 1.0f,
                screenPos.z * 2.0f - 1.0f);
        
        // Transform to world position.
        out.transformMat4(out, _matInvViewProj);
    }
    
    return out;
}

Vec3& Camera::worldToScreen(Vec3& out, const Vec3& worldPos, int width, int height)
{
    calcMatrices(width, height);
    
    float cx = _rect.x * width;
    float cy = _rect.y * height;
    float cw = _rect.w * width;
    float ch = _rect.h * height;
    
    out.transformMat4(worldPos, _matViewProj);
    out.x = cx + (out.x + 1) * 0.5f * cw;
    out.y = cy + (out.y + 1) * 0.5f * ch;
    out.z = out.z * 0.5 + 0.5;
    
    return out;
}

Mat4& Camera::worldMatrixToScreen(Mat4& out, const Mat4& worldMatrix, int width, int height)
{
    calcMatrices(width, height);
    
    Mat4::multiply(_matViewProj, worldMatrix, &out);

    float halfWidth = width / 2;
    float halfHeight = height / 2;
    
    _temp_mat4.set(Mat4::IDENTITY);
    _temp_mat4.translate(halfWidth, halfHeight, 0, &_temp_mat4);
    _temp_mat4.scale(halfWidth, halfHeight, 1, &_temp_mat4);

    out.multiply(_temp_mat4);

    return out;
}

RENDERER_END

File: cocos2d/renderer/scene/camera.js
// Copyright (c) 2017-2018 Xiamen Yaji Software Co., Ltd.

import { Vec3, Mat4, lerp, Vec4 } from '../../core/value-types';
import { Ray } from '../../core/geom-utils';
import enums from '../enums';

let _tmp_mat4 = new Mat4();

let _matView = new Mat4();
let _matViewInv = new Mat4();
let _matProj = new Mat4();
let _matViewProj = new Mat4();
let _matInvViewProj = new Mat4();
let _tmp_v3 = new Vec3();
let _tmp2_v3 = new Vec3();

/**
 * A representation of a camera instance
 */
export default class Camera {
  _poolID = -1;
  _node = null;
  _projection = enums.PROJ_PERSPECTIVE;

  // priority. the smaller one will be rendered first
  _priority = 0;

  // clear options
  _color = new Vec4(0.2, 0.3, 0.47, 1);
  _depth = 1;
  _stencil = 0;
  _clearFlags = enums.CLEAR_COLOR | enums.CLEAR_DEPTH;
  _clearModel = null;

  // stages & framebuffer
  _stages = [];
  _framebuffer = null;

  // projection properties
  _near = 0.01;
  _far = 1000.0;
  _fov = Math.PI / 4.0; // vertical fov
  _rect = {
    x: 0, y: 0, w: 1, h: 1
  };

  // ortho properties
  _orthoHeight = 10;

  _cullingMask = 0xffffffff;


  // culling mask
  get cullingMask () {
    return this._cullingMask;
  }

  set cullingMask (mask) {
    this._cullingMask = mask;
  }

  setCullingMask (mask) {
    this._cullingMask = mask;
  }

  /**
   * Get the hosting node of this camera
   * @returns {Node} the hosting node
   */
  getNode () {
    return this._node;
  }

  /**
   * Set the hosting node of this camera
   * @param {Node} node the hosting node
   */
  setNode (node) {
    this._node = node;
  }

  /**
   * Get the projection type of the camera
   * @returns {number} camera projection type
   */
  getType () {
    return this._projection;
  }

  /**
   * Set the projection type of the camera
   * @param {number} type camera projection type
   */
  setType (type) {
    this._projection = type;
  }

  /**
   * Get the priority of the camera
   * @returns {number} camera priority
   */
  getPriority () {
    return this._priority;
  }

  /**
   * Set the priority of the camera
   * @param {number} priority camera priority
   */
  setPriority (priority) {
    this._priority = priority;
  }

  /**
   * Get the orthogonal height of the camera
   * @returns {number} camera height
   */
  getOrthoHeight () {
    return this._orthoHeight;
  }

  /**
   * Set the orthogonal height of the camera
   * @param {number} val camera height
   */
  setOrthoHeight (val) {
    this._orthoHeight = val;
  }

  /**
   * Get the field of view of the camera
   * @returns {number} camera field of view
   */
  getFov () {
    return this._fov;
  }

  /**
   * Set the field of view of the camera
   * @param {number} fov camera field of view
   */
  setFov (fov) {
    this._fov = fov;
  }

  /**
   * Get the near clipping distance of the camera
   * @returns {number} camera near clipping distance
   */
  getNear () {
    return this._near;
  }

  /**
   * Set the near clipping distance of the camera
   * @param {number} near camera near clipping distance
   */
  setNear (near) {
    this._near = near;
  }

  /**
   * Get the far clipping distance of the camera
   * @returns {number} camera far clipping distance
   */
  getFar () {
    return this._far;
  }

  /**
   * Set the far clipping distance of the camera
   * @param {number} far camera far clipping distance
   */
  setFar (far) {
    this._far = far;
  }

  /**
   * Get the clear color of the camera
   * @returns {Vec4} out the receiving color vector
   */
  getColor (out) {
    return Vec4.copy(out, this._color);
  }

  /**
   * Set the clear color of the camera
   * @param {number} r red channel of camera clear color
   * @param {number} g green channel of camera clear color
   * @param {number} b blue channel of camera clear color
   * @param {number} a alpha channel of camera clear color
   */
  setColor (r, g, b, a) {
    Vec4.set(this._color, r, g, b, a);
  }

  /**
   * Get the clear depth of the camera
   * @returns {number} camera clear depth
   */
  getDepth () {
    return this._depth;
  }

  /**
   * Set the clear depth of the camera
   * @param {number} depth camera clear depth
   */
  setDepth (depth) {
    this._depth = depth;
  }

  /**
   * Get the clearing stencil value of the camera
   * @returns {number} camera clearing stencil value
   */
  getStencil () {
    return this._stencil;
  }

  /**
   * Set the clearing stencil value of the camera
   * @param {number} stencil camera clearing stencil value
   */
  setStencil (stencil) {
    this._stencil = stencil;
  }

  /**
   * Get the clearing flags of the camera
   * @returns {number} camera clearing flags
   */
  getClearFlags () {
    return this._clearFlags;
  }

  /**
   * Set the clearing flags of the camera
   * @param {number} flags camera clearing flags
   */
  setClearFlags (flags) {
    this._clearFlags = flags;
  }

  /**
   * Get the rect of the camera
   * @param {Object} out the receiving object
   * @returns {Object} camera rect
   */
  getRect (out) {
    out.x = this._rect.x;
    out.y = this._rect.y;
    out.w = this._rect.w;
    out.h = this._rect.h;

    return out;
  }

  /**
   * Set the rect of the camera
   * @param {Number} x - [0,1]
   * @param {Number} y - [0,1]
   * @param {Number} w - [0,1]
   * @param {Number} h - [0,1]
   */
  setRect (x, y, w, h) {
    this._rect.x = x;
    this._rect.y = y;
    this._rect.w = w;
    this._rect.h = h;
  }

  /**
   * Get the stages of the camera
   * @returns {string[]} camera stages
   */
  getStages () {
    return this._stages;
  }

  /**
   * Set the stages of the camera
   * @param {string[]} stages camera stages
   */
  setStages (stages) {
    this._stages = stages;
  }

  /**
   * Get the framebuffer of the camera
   * @returns {FrameBuffer} camera framebuffer
   */
  getFramebuffer () {
    return this._framebuffer;
  }

  /**
   * Set the framebuffer of the camera
   * @param {FrameBuffer} framebuffer camera framebuffer
   */
  setFrameBuffer (framebuffer) {
    this._framebuffer = framebuffer;
  }

  _calcMatrices (width, height) {
    // view matrix
    this._node.getWorldRT(_matViewInv);
    Mat4.invert(_matView, _matViewInv);

    // projection matrix
    let aspect = width / height;
    if (this._projection === enums.PROJ_PERSPECTIVE) {
      Mat4.perspective(_matProj,
        this._fov,
        aspect,
        this._near,
        this._far
      );
    } else {
      let x = this._orthoHeight * aspect;
      let y = this._orthoHeight;
      Mat4.ortho(_matProj,
        -x, x, -y, y, this._near, this._far
      );
    }

    // view-projection
    Mat4.mul(_matViewProj, _matProj, _matView);
    // inv view-projection
    Mat4.invert(_matInvViewProj, _matViewProj);
  }

  /**
   * extract a view of this camera
   * @param {View} out the receiving view
   * @param {number} width framebuffer width
   * @param {number} height framebuffer height
   */
  extractView (out, width, height) {
    if (this._framebuffer) {
      width = this._framebuffer._width;
      height = this._framebuffer._height;
    }

    // priority
    out._priority = this._priority;

    // rect
    out._rect.x = this._rect.x * width;
    out._rect.y = this._rect.y * height;
    out._rect.w = this._rect.w * width;
    out._rect.h = this._rect.h * height;

    // clear opts
    this.getColor(out._color);
    out._depth = this._depth;
    out._stencil = this._stencil;
    out._clearFlags = this._clearFlags;
    out._clearModel = this._clearModel;

    // stages & framebuffer
    out._stages = this._stages;
    out._framebuffer = this._framebuffer;

    this._calcMatrices(width, height);
    Mat4.copy(out._matView, _matView);
    Mat4.copy(out._matViewInv, _matViewInv);
    Mat4.copy(out._matProj, _matProj);
    Mat4.copy(out._matViewProj, _matViewProj);
    Mat4.copy(out._matInvViewProj, _matInvViewProj);

    out._cullingMask = this._cullingMask;
  }

  /**
   * transform a screen position to a world space ray
   * @param {number} x the screen x position to be transformed
   * @param {number} y the screen y position to be transformed
   * @param {number} width framebuffer width
   * @param {number} height framebuffer height
   * @param {Ray} out the resulting ray
   * @returns {Ray} the resulting ray
   */
  screenPointToRay (x, y, width, height, out) {
    if (!cc.geomUtils) return out;

    out = out || new Ray();
    this._calcMatrices(width, height);

    let cx = this._rect.x * width;
    let cy = this._rect.y * height;
    let cw = this._rect.w * width;
    let ch = this._rect.h * height;

    // far plane intersection
    Vec3.set(_tmp2_v3, (x - cx) / cw * 2 - 1, (y - cy) / ch * 2 - 1, 1);
    Vec3.transformMat4(_tmp2_v3, _tmp2_v3, _matInvViewProj);

    if (this._projection === enums.PROJ_PERSPECTIVE) {
      // camera origin
      this._node.getWorldPosition(_tmp_v3);
    } else {
      // near plane intersection
      Vec3.set(_tmp_v3, (x - cx) / cw * 2 - 1, (y - cy) / ch * 2 - 1, -1);
      Vec3.transformMat4(_tmp_v3, _tmp_v3, _matInvViewProj);
    }

    return Ray.fromPoints(out, _tmp_v3, _tmp2_v3);
  }

  /**
   * transform a screen position to world space
   * @param {Vec3} out the resulting vector
   * @param {Vec3} screenPos the screen position to be transformed
   * @param {number} width framebuffer width
   * @param {number} height framebuffer height
   * @returns {Vec3} the resulting vector
   */
  screenToWorld (out, screenPos, width, height) {
    this._calcMatrices(width, height);

    let cx = this._rect.x * width;
    let cy = this._rect.y * height;
    let cw = this._rect.w * width;
    let ch = this._rect.h * height;

    if (this._projection === enums.PROJ_PERSPECTIVE) {
      // calculate screen pos in far clip plane
      Vec3.set(out,
        (screenPos.x - cx) / cw * 2 - 1,
        (screenPos.y - cy) / ch * 2 - 1,
        0.9999
      );

      // transform to world
      Vec3.transformMat4(out, out, _matInvViewProj);

      // lerp to depth z
      this._node.getWorldPosition(_tmp_v3);

      Vec3.lerp(out, _tmp_v3, out, lerp(this._near / this._far, 1, screenPos.z));
    } else {
      Vec3.set(out,
        (screenPos.x - cx) / cw * 2 - 1,
        (screenPos.y - cy) / ch * 2 - 1,
        screenPos.z * 2 - 1
      );

      // transform to world
      Vec3.transformMat4(out, out, _matInvViewProj);
    }

    return out;
  }

  /**
   * transform a world space position to screen space
   * @param {Vec3} out the resulting vector
   * @param {Vec3} worldPos the world space position to be transformed
   * @param {number} width framebuffer width
   * @param {number} height framebuffer height
   * @returns {Vec3} the resulting vector
   */
  worldToScreen (out, worldPos, width, height) {
    this._calcMatrices(width, height);

    let cx = this._rect.x * width;
    let cy = this._rect.y * height;
    let cw = this._rect.w * width;
    let ch = this._rect.h * height;

    Vec3.transformMat4(out, worldPos, _matViewProj);
    out.x = cx + (out.x + 1) * 0.5 * cw;
    out.y = cy + (out.y + 1) * 0.5 * ch;
    out.z = out.z * 0.5 + 0.5;

    return out;
  }

  /**
   * transform a world space matrix to screen space
   * @param {Mat4} out the resulting vector
   * @param {Mat4} worldMatrix the world space matrix to be transformed
   * @param {number} width framebuffer width
   * @param {number} height framebuffer height
   * @returns {Mat4} the resulting vector
   */
  worldMatrixToScreen (out, worldMatrix, width, height) {
    this._calcMatrices(width, height);

    Mat4.mul(out, _matViewProj, worldMatrix);

    let halfWidth = width / 2;
    let halfHeight = height / 2;
    Mat4.identity(_tmp_mat4);
    Mat4.transform(_tmp_mat4, _tmp_mat4, Vec3.set(_tmp_v3, halfWidth, halfHeight, 0));
    Mat4.scale(_tmp_mat4, _tmp_mat4, Vec3.set(_tmp_v3, halfWidth, halfHeight, 1));

    Mat4.mul(out, _tmp_mat4, out);

    return out;
  }
}
